server/src/main/java/org/elasticsearch/plugins/UberModuleClassLoader.java

```diff
@@ -0,0 +1,281 @@
+            }
+            byte[] bytes = in.readAllBytes();
+            return defineClass(name, bytes, 0, bytes.length, codeSource);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    /**
+     * This classloader does not restrict access to resources in its jars. Users should
+     * expect the same behavior as that provided by {@link URLClassLoader}.
+     *
+     * @param moduleName Name of this classloader's synthetic module
+     * @param name The resource name
+     * @return a URL for the resource, or null if the resource could not be found,
+     *   if the module name does not match, or if the loader is closed.
+     */
+    @Override
+    protected URL findResource(String moduleName, String name) {
+        if (Objects.isNull(moduleName) || this.module.getName().equals(moduleName) == false) {
+            return null;
+        }
+        return findResource(name);
+    }
+
+    /**
+     * This classloader does not restrict access to resources in its jars. Users should
+     * expect the same behavior as that provided by {@link URLClassLoader}.
+     *
+     * @param name The resource name
+     * @return a URL for the resource, or null if the resource could not be found,
+     *   or if the loader is closed.
+     */
+    @Override
+    protected URL findResource(String name) {
+        return internalLoader.findResource(name);
+    }
+
+    /**
+     * This classloader does not restrict access to resources in its jars. Users should
+     * expect the same behavior as that provided by {@link URLClassLoader}.
+     *
+     * @param name The resource name
+     * @return an Enumeration of URLs. If the loader is closed, the Enumeration contains no elements.
+     */
+    @Override
+    protected Enumeration<URL> findResources(String name) throws IOException {
+        return internalLoader.findResources(name);
+    }
+
+    @Override
+    protected Class<?> loadClass(String cn, boolean resolve) throws ClassNotFoundException {
+        // if cn's package is in this ubermodule, look here only (or just first?)
+        String packageName = packageName(cn);
+
+        // TODO: do we need to check the security manager here?
+
+        synchronized (getClassLoadingLock(cn)) {
+
+            // check if already loaded
+            Class<?> c = findLoadedClass(cn);
+
+            if (c == null) {
+                if (this.packageNames.contains(packageName)) {
+                    // find in module or null
+                    c = findClass(cn);
+                } else {
+                    c = getParent().loadClass(cn);
+                }
+            }
+
+            if (c == null) {
+                throw new ClassNotFoundException(cn);
+            }
+
+            if (resolve) {
+                resolveClass(c);
+            }
+
+            return c;
+        }
+    }
+
+    /**
+     * Returns the package name for the given class name
+     * (from jdk.internal.loader.Loader)
```

Comment:
we could remove this comment (referring to jdk.internal), as the implementation is trivial.